---
title: 'TCGA Workflow: Analyze cancer genomics and epigenomics data using Bioconductor
  packages'
author: Tiago C. Silva, Antonio Colaprico, Catharina Olsen, Fulvio Dâ€™Angelo, Gianluca
  Bontempi Michele Ceccarelli , and Houtan Noushmehr
date: "`r Sys.Date()`"
vignette: >
  %\VignetteIndexEntry{"Introduction"}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: bibliography.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(progress = FALSE)
```
```{r, echo=FALSE, results="hide", warning=FALSE,message=FALSE}
library(BiocStyle)
suppressPackageStartupMessages({
  if("TCGAWorkflow" %in% installed.packages()[,1]) {
    library(TCGAWorkflow)
  } else {
    devtools::load_all(".")
  }
})
```


# About

This workflow is based on the article: [TCGA Workflow: Analyze cancer genomics and epigenomics data using Bioconductor packages](https://f1000research.com/articles/5-1542/v2) [@10.12688/f1000research.8923.2].
Due to time and space contraints we downloaded only a subset of the data, for a real analysis please use all data available. The data used in the examples are available in the package `r Biocpkg("TCGAWorkflowData")`. 

## Installation

To be able to execute all the steps of this workflow please install it with the following code:

```{R, eval=FALSE, include=TRUE}
source("https://bioconductor.org/biocLite.R")
deps <- c("pathview","clusterProfiler","ELMER", "DO.db","GO.db", 
          "ComplexHeatmap","EDASeq", "TCGAbiolinks","AnnotationHub",
          "gaia","ChIPseeker","minet","BSgenome.Hsapiens.UCSC.hg19",
          "MotifDb","MotIV", "rGADEM", "motifStack","RTCGAToolbox")
for(pkg in deps)  if (!pkg %in% installed.packages()) biocLite(pkg, dependencies = TRUE)
deps <- c("devtools","DT","pbapply","readr","circlize")
for(pkg in deps)  if (!pkg %in% installed.packages())  install.packages(pkg,dependencies = TRUE)
devtools::install_github("BioinformaticsFMRP/TCGAWorkflowData")
devtools::install_github("BioinformaticsFMRP/TCGAWorkflow", dependencies = TRUE)
```

## Loading packages

At the beginning of each section, the packages required to execute the code will be loaded. However the following packages are required for all sections.

- TCGAWorkflowData: this package contains the data necessary to execute each of the analysis
steps. This is a subset of the downloaded to make the example faster. For a real analysis, please
use all the data available.
- DT: we will use it to visualize the results

```{R, eval=TRUE, message=FALSE, warning=FALSE, include=TRUE}
library(TCGAWorkflowData)
library(DT)
```

# Abstract

Biotechnological advances in sequencing have led to an explosion of
    publicly available data via large international consortia such as [The
    Cancer Genome Atlas (TCGA)](http://cancergenome.nih.gov/), [The
    Encyclopedia of DNA Elements (ENCODE)](http://www.encodeproject.org/),
    and [The NIH Roadmap Epigenomics Mapping Consortium
    (Roadmap)](http://www.roadmapepigenomics.org/). These projects have
    provided unprecedented opportunities to interrogate the epigenome of
    cultured cancer cell lines as well as normal and tumor tissues with high
    genomic resolution. The [Bioconductor](http://www.bioconductor.org/)
    project offers more than 1,000 open-source software and statistical
    packages to analyze high-throughput genomic data. However, most packages
    are designed for specific data types (e.g. expression, epigenetics,
    genomics) and there is no one comprehensive tool that provides a
    complete integrative analysis of the resources and data provided by all
    three public projects. A need to create an integration of these
    different analyses was recently proposed. In this workflow, we provide a
    series of biologically focused integrative analyses of different
    molecular data. We describe how to download, process and prepare TCGA
    data and by harnessing several key Bioconductor packages, we describe
    how to extract biologically meaningful genomic and epigenomic data.
    Using Roadmap and ENCODE data, we provide a work plan to identify
    biologically relevant functional epigenomic elements associated with
    cancer. 
    
To illustrate our workflow, we analyzed two types of brain
tumors: low-grade glioma (LGG) versus high-grade glioma (glioblastoma
multiform or GBM). 

All the  package landing pages used in this workflow can be found through the [biocViews interface](http://www.bioconductor.org/packages/release/BiocViews.html#___Software).
    
**Keywords:** Epigenomics, Genomics, Cancer, non-coding, TCGA, ENCODE, Roadmap, Bioinformatics.

# Introduction

Cancer is a complex genetic disease spanning multiple molecular events
such as point mutations, structural variations, translocations and
activation of epigenetic and transcriptional signatures and networks.
The effects of these events take place at different spatial and temporal
scales with interlayer communications and feedback mechanisms creating a
highly complex dynamic system. To gain insight into the biology of
tumours most of the research in cancer genomics is aimed at the
integration of the observations at multiple molecular scales and the
analysis of their interplay. Even if many tumors share similar recurrent
genomic events, their relationships with the observed phenotype are
often not understood. For example, although we know that the majority of
the most aggressive form of brain tumours such as glioma harbour the
mutation of a single gene (IDH), the mechanistic explanation of the
activation of its characteristic epigenetic and transcriptional
signatures are still far to be well characterized. Moreover,
network-based strategies have recently emerged as an effective framework
for the discovery functional disease drivers that act as main regulators
of cancer phenotypes. Here we describe a comprehensice workflow that
integrates many Bioconductor packages in order to analyze and integrate
the molteplicity of molecular observation layers in large scale cancer
dataset.

Indeed, recent technological developments allowed the deposition of
large amounts of genomic and epigenomic data, such as gene expression,
DNA methylation, and genomic localization of transcription factors, into
freely available public international consortia like The Cancer Genome
Atlas ([TCGA](http://cancergenome.nih.gov/)), The Encyclopedia of DNA
Elements ([ENCODE](http://www.encodeproject.org/)), and The NIH Roadmap
Epigenomics Mapping Consortium
([Roadmap](http://www.roadmapepigenomics.org/)) [@Hawkins]. An overview
of the three consortia is described below:

-   **The Cancer Genome Atlas (TCGA):** The TCGA consortium, which is a
    National Institute of Health (NIH) initiative, makes publicly
    available molecular and clinical information for more than 30 types
    of human cancers including exome (variant analysis), single
    nucleotide polymorphism (SNP), DNA methylation, transcriptome
    (mRNA), microRNA (miRNA) and proteome . Sample types available at
    TCGA are: primary solid tumors, recurrent solid tumors, blood
    derived normal and tumor, metastatic, and solid tissue normal
    [@weinstein2013cancer].

-   **The Encyclopedia of DNA Elements (ENCODE):** Found in 2003 by the
    National Human Genome Research Institute (NHGRI), the project aims
    to build a comprehensive list of functional elements that have an
    active role in the genome, including regulatory elements that govern
    gene expression. Biosamples includes immortalized cell lines,
    tissues, primary cells and stem cells [@encode2011user].

-   **The NIH Roadmap Epigenomics Mapping Consortium:** This was
    launched with the goal of producing a public resource of human
    epigenomic data in order to analyze biology and disease-oriented
    research. Roadmap maps DNA methylation, histone modifications,
    chromatin accessibility, and small RNA transcripts in stem cells and
    primary ex vivo tissues [@Fingerman; @Bernstein].

Briefly, these three consortia provide large scale epigenomic data onto
a variety of microarrays and next-generation sequencing (NGS) platforms.
Each consortium encompasses specific types of biological information on
specific type of tissue or cell and when analyzed together, it provides
an invaluable opportunity for research laboratories to better understand
the developmental progression of normal cells to cancer state at the
molecular level and importantly, correlate these phenotypes with tissue
of origins.

Although there exists a wealth of possibilities [@kannan2015public] in
accessing cancer associated data,
[Bioconductor](http://www.bioconductor.org/) represent the most
comprehensive set of open source, updated and integrated professional
tools for the statistical analysis of large scale genomic data. Thus, we
propose our workflow within Bioconductor to describe how to download,
process, analyze and integrate cancer data to understand specific
cancer-related specific questions. However, there is no tool that solves
the issue of integration in a comprehensive sequence and mutation
information, epigenomic state and gene expression within the context of
gene regulatory networks to identify oncogenic drivers and characterize
altered pathways during cancer progression. Therefore, our workflow
presents several [Bioconductor](http://www.bioconductor.org/) packages
to work with genomic and epigenomics data.

# References
